---
title: "Overlap analyses for all the Boxer et al Dataset"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\VignetteDepends{devtools}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
Set up dependencies and load package
```{r setup}
library(OverlapPlots)

```

```{r, include=FALSE}
library(devtools)
devtools::load_all(".")
```

INFO SESSION:

```{r}
library(details)
library(sessioninfo)

sessioninfo::session_info()%>%
     details::details(
         summary = 'Current session info',
         open    = TRUE
     )
```

Set directory

```{r}

# Set up working directory
rm(list = ls())
#setwd("C:/Users/danie/Documents/longene2/R")
print(getwd())

```
Load annotation dataset reference sequence gene
```{r}
load(file = "../dat-info/mm10_ncbi-refSeqGene_Dec2019.RData")
genotypes <- factor(c(rep("WT", 10), rep("KO", 10)), levels = c("KO", "WT"))

```
overlap plots -- mean(log2FC) vs mean(gene length)
```{r}
## KO/WT whole cell dataset (ref: Fig 1D Boxer et al. Mol Cell 2020)
count.file <- "../dat/counts/GSE128178_10WT_10MeCP2_KO_whole_cell_RNAseq_exon_counts.txt"
degs.file <- "../dat/DEGs/KO-WT_whole-cell_RNA-seq.txt"
bin.size <- 60
shift.size <- 6
wholeCell.KO <- finalRun(count.file, genotypes, degs.file, bin.size, shift.size,
                         title = "KO/WT whole cell dataset")

## KO/WT nuclear dataset  (ref: Fig 2E Boxer et al. Mol Cell 2020)
count.file <- "../dat/counts/GSE128178_10WT_10MeCP2_KO_nuclear_RNAseq_gene_body_counts.txt"
degs.file <- "../dat/DEGs/KO-WT_nuclear_RNA-seq.txt"
bin.size <- 40
shift.size <- 4
nuclear.KO <- finalRun(count.file, genotypes, degs.file, bin.size, shift.size,
                       title = "KO/WT nuclear dataset")

## KO/WT chromatin dataset (ref: Fig 2F Boxer et al. Mol Cell 2020)
count.file <- "../dat/counts/GSE128178_10WT_10MeCP2_KO_chromatin_associated_RNAseq_gene_body_counts.txt"
degs.file <- "../dat/DEGs/KO-WT_chromatin_RNA-seq.txt"
bin.size <- 40
shift.size <- 4
chromatin.KO <- finalRun(count.file, genotypes, degs.file, bin.size, shift.size,
                         title = "KO/WT chromatin dataset")

## R306C/WT whole cell dataset  (ref: Fig 3F Boxer et al. Mol Cell 2020)
count.file <- "../dat/counts/GSE128178_10WT_10R306C_whole_cell_RNAseq_exon_counts.txt"
degs.file <- "../dat/DEGs/R306C-WT_whole-cell_RNA-seq.txt"
bin.size <- 40
shift.size <- 4
wholeCell.R306C <- finalRun(count.file, genotypes, degs.file, bin.size, 
                            shift.size, title = "R306C/WT whole cell dataset")

## R306C/WT nuclear dataset 
count.file <- "../dat/counts/GSE128178_10WT_10R306C_nuclear_RNAseq_gene_body_counts.txt"
degs.file <- "../dat/DEGs/R306C-WT_nuclear_RNA-seq.txt"
bin.size <- 40
shift.size <- 4
nuclear.R306C <- finalRun(count.file, genotypes, degs.file, bin.size, 
                          shift.size, title = "R306C/WT nuclear dataset")

## R306C/WT chromatin dataset
count.file <- "../dat/counts/GSE128178_10WT_10R306C_chromatin_associated_RNAseq_gene_body_counts.txt"
degs.file <- "../dat/DEGs/R306C-WT_chromatin_RNA-seq.txt"
bin.size <- 40
shift.size <- 4
chromatin.R306C <- finalRun(count.file, genotypes, degs.file, bin.size, 
                            shift.size, title = "R306C/WT chromatin dataset")
```
overlap plots -- mean(log2FC) vs mean(mCA)
```{r}
## load the workspace
load(file = "../dat-info/mm10_ncbi-refSeqGene_Dec2019.RData")
mCA <- data.frame(readRDS("../dat/mCA/CAperGene_10wk_boxer.RDS"), stringsAsFactors = FALSE)
mCA <- mCA[mCA$CA >= 5,]
mCA <- mCA[mCA$gene.length >= 4500,]
mCA$mCA.CA <- mCA$mCA/mCA$CA
mCA1 <- mCA[,c(1,6,7)]

## mCA analysis for KO/WT in whole cell dataset
degs.dat <- read.table("../dat/DEGs/KO-WT_whole-cell_RNA-seq.txt", sep = "\t", 
                       stringsAsFactors = FALSE, header = TRUE, row.names = 1)
mat <- wholeCell.KO$res$results[,c(8:27,1,3)]
colnames(mat)[21] <- "gene.name"
grp.idx <- WTgrpKmeans(control_mat = mat[,1:10])
if(length(grp.idx$WT.idx1) != length(grp.idx$WT.idx2)){
    message("Cluster size is not equal, therefore run same size k-means 
                variation!")
    grp.idx <- WTgrpKmeansEqualSize(control_mat = mat[,1:10])
}
res1 <- overlapDegsmCAWrapper(degs.dat = degs.dat, count.dat = mat,
                                 refseq = mCA1, WT1.idx = grp.idx$WT.idx1, 
                                 WT2.idx = grp.idx$WT.idx2, bin.size = 60, 
                                 shift.size = 6, methyl.type = "mCA/CA")

p1 <- (res1$plot1 + coord_cartesian(ylim = c(-0.25,0.25))) / res1$plot2 / res1$plot4
rm(wholeCell.KO, degs.dat, mat)

## mCA analysis for R306C/WT
degs.dat <- read.table("../dat/DEGs/R306C-WT_whole-cell_RNA-seq.txt", 
                       sep = "\t", stringsAsFactors = FALSE, header = TRUE, 
                       row.names = 1)
mat <- wholeCell.R306C$res$results[,c(8:27,1,3)]
colnames(mat)[21] <- "gene.name"
grp.idx <- WTgrpKmeans(control_mat = mat[,1:10])
if(length(grp.idx$WT.idx1) != length(grp.idx$WT.idx2)){
    message("Cluster size is not equal, therefore run same size k-means 
                variation!")
    grp.idx <- WTgrpKmeansEqualSize(control_mat = mat[,1:10])
}
res2 <- overlapDegsmCAWrapper(degs.dat, count.dat = mat, refseq = mCA1, 
                                 WT1.idx = grp.idx$WT.idx1, 
                                 WT2.idx = grp.idx$WT.idx2, bin.size = 40, 
                                 shift.size = 4, methyl.type = "mCA/CA")
p2 <- (res2$plot1 + coord_cartesian(ylim = c(-0.25,0.25))) / res2$plot2 / res2$plot4
rm(wholeCell.R306C, degs.dat)

```
