---
title: "Overlap analyses for all the Boxer et al Dataset"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{my-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\VignetteDepends{devtools}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
Set up dependencies and load package
```{r setup}
library(OverlapPlots)

```

```{r, include=FALSE}
library(devtools)
devtools::load_all(".")
```

INFO SESSION:

```{r}
library(details)
library(sessioninfo)

sessioninfo::session_info()%>%
     details::details(
         summary = 'Current session info',
         open    = TRUE
     )
```

Overlap analyses for all the Boxer et al Dataset

```{r}

# Set up working directory
rm(list = ls())
#setwd("C:/Users/danie/Documents/longene2/R")
print(getwd())

```
Load annotation dataset reference sequence gene
```{r}
load(file = "../dat-info/mm10_ncbi-refSeqGene_Dec2019.RData")
genotypes <- factor(c(rep("WT", 10), rep("KO", 10)), levels = c("KO", "WT"))

```

Load KO/WT whole cell dataset

```{r}

dat <- read.table("../dat/counts/GSE128178_10WT_10MeCP2_KO_whole_cell_RNAseq_exon_counts.txt",
                  sep = "\t", stringsAsFactors=FALSE, header = TRUE,
                  row.names = 1)

```
Run Differential Expression Analysis
```{r}
wholeCell.KO <- DESeqCalculation(dat = dat, genotypes = genotypes, fc = 1.15)
mat <- wholeCell.KO$results[,c(8:27,1,3)]
colnames(mat)[21] <- "gene.name"

```
Run k-means clustering to obtained the different WT groups
```{r}
grp.idx <- WTgrp_kmeans(control_mat = mat[,1:10])

```
Plot running average figure with wrapper function
```{r}
res1 <- overlap_wrapper(dat = mat, refseq = refseq, KO.idx = c(11:20),
                        WT.idx = c(1:10), WT1.idx = grp.idx$WT.idx1,
                        WT2.idx = grp.idx$WT.idx2, bin.size = 200,
                        shift.size = 40)
res1$plot

```

For the second plot we consider the DEGs list. First, we load the dataset 
and filter for FDR less than 0.05

```{r}
degs.dat <- read.table("../dat/DEGs/KO-WT_whole-cell_RNA-seq.txt", sep = "\t",
                       stringsAsFactors=FALSE, header = TRUE, row.names = 1)
degs.dat <- degs.dat[degs.dat$FDR < 0.05,]

cat("Number of DEGs =",dim(degs.dat)[1],"\n\n")
degs.dat <- inner_join(degs.dat %>% rownames_to_column(var = "gene.name"),
                       refseq, by = "gene.name")

```
Generate scatter plots of log fold change vs gene length.
```{r}
Plot.Scatter(dat = degs.dat[,c("gene.name","logFC", "FDR", "gene.length")],
             log2FC = log2(1), comp.between = "")
Plot.Scatter(dat = degs.dat[,c("gene.name","logFC", "FDR", "gene.length")],
             log2FC = log2(1.2), comp.between = "")
degs.dat <- inner_join(x = wholeCell.KO$results[wholeCell.KO$results$gene %in%
                                                degs.dat$gene.name,c(8:27,1,3)],
                       y = degs.dat[,c("gene.name", "logFC")],
                       by = c("gene" = "gene.name"))
mat <- degs.dat[,c(1:21,23)]
colnames(mat)[21] <- "gene.name"

```
logFC from Degs from edgeR (Boxer et al.)
```{r}
colnames(mat)[22] <- "log2FoldChange"
res2 <- overlap_wrapper(dat = mat, refseq = refseq, KO.idx = c(11:20),
                        WT.idx = c(1:10), WT1.idx = grp.idx$WT.idx1,
                        WT2.idx = grp.idx$WT.idx2, bin.size = 60,
                        shift.size = 6, shrink_lfc = TRUE)
res2$plot
```

logFC from Degs from DESeq2
```{r}
mat <- degs.dat[,c(1:22)]
colnames(mat)[21] <- "gene.name"
res3 <- overlap_wrapper(dat = mat, refseq = refseq, KO.idx = c(11:20),
                        WT.idx = c(1:10), WT1.idx = grp.idx$WT.idx1,
                        WT2.idx = grp.idx$WT.idx2, bin.size = 60,
                        shift.size = 6, shrink_lfc = TRUE)
res3$plot

```
Load KO/WT nuclear dataset
```{r}
rm(dat, wholeCell.KO, mat, grp.idx, degs.dat)
dat <- read.table("../dat/counts/GSE128178_10WT_10MeCP2_KO_nuclear_RNAseq_gene_body_counts.txt",
                  sep = "\t", stringsAsFactors=FALSE, header = TRUE,
                  row.names = 1)

```
Run Differential Expresion analysis.
```{r}
nuclear.KO <- DESeqCalculation(dat = dat, genotypes = genotypes, fc = 1.15)
mat <- nuclear.KO$results[,c(8:27,1,3)]
colnames(mat)[21] <- "gene.name"

```
Run k-means clustering on control (WT)
```{r}
grp.idx <- WTgrp_kmeans(control_mat = mat[,1:10])
if(length(grp.idx$WT.idx1) != length(grp.idx$WT.idx2)){
message("Cluster size is not equal, therefore run same size k-means variation!")
    grp.idx <- WTgrp_kmeans_eqSize(control_mat = mat[,1:10])
}
```
Plot running average figure with wrapper function
```{r}
res4 <- overlap_wrapper(dat = mat, refseq = refseq, KO.idx = c(11:20),
                        WT.idx = c(1:10), WT1.idx = grp.idx$WT.idx1,
                        WT2.idx = grp.idx$WT.idx2, bin.size = 200,
                        shift.size = 40)
res4$plot

```
Now considering DEGs list
```{r}
degs.dat <- read.table("../dat/DEGs/KO-WT_nuclear_RNA-seq.txt", sep = "\t",
                       stringsAsFactors=FALSE, header = TRUE, row.names = 1)
degs.dat <- degs.dat[degs.dat$FDR < 0.05,]
cat("Number of DEGs =",dim(degs.dat)[1],"\n\n")
degs.dat <- inner_join(degs.dat %>% rownames_to_column(var = "gene.name"),
                       refseq, by = "gene.name")

```
Generate scatter plots of log fold change vs gene length.
```{r}
Plot.Scatter(dat = degs.dat[,c("gene.name","logFC", "FDR", "gene.length")],
             log2FC = log2(1), comp.between = "")
Plot.Scatter(dat = degs.dat[,c("gene.name","logFC", "FDR", "gene.length")],
             log2FC = log2(1.2), comp.between = "")
degs.dat <- inner_join(x = nuclear.KO$results[nuclear.KO$results$gene %in%
                                              degs.dat$gene.name,c(8:27,1,3)],
                       y = degs.dat[,c("gene.name", "logFC")],
                       by = c("gene" = "gene.name"))
mat <- degs.dat[,c(1:21,23)]
colnames(mat)[21] <- "gene.name"

```
logFC from Degs from edgeR (Boxer et al.)
```{r}
colnames(mat)[22] <- "log2FoldChange"
res5 <- overlap_wrapper(dat = mat, refseq = refseq, KO.idx = c(11:20),
                        WT.idx = c(1:10), WT1.idx = grp.idx$WT.idx1,
                        WT2.idx = grp.idx$WT.idx2, bin.size = 40,
                        shift.size = 4, shrink_lfc = TRUE)
res5$plot
```
logFC from Degs from DESeq2
```{r}
mat <- degs.dat[,c(1:22)]
colnames(mat)[21] <- "gene.name"
res6 <- overlap_wrapper(dat = mat, refseq = refseq, KO.idx = c(11:20),
                        WT.idx = c(1:10), WT1.idx = grp.idx$WT.idx1,
                        WT2.idx = grp.idx$WT.idx2, bin.size = 40,
                        shift.size = 4, shrink_lfc = TRUE)
res6$plot
rm(dat, nuclear.KO, mat, grp.idx, degs.dat)

```
Load KO/WT chromatin dataset
```{r}
dat <- read.table("../dat/counts/GSE128178_10WT_10MeCP2_KO_chromatin_associated_RNAseq_gene_body_counts.txt",
              sep = "\t", stringsAsFactors=FALSE, header = TRUE, row.names = 1)
chromatin.KO <- DESeqCalculation(dat = dat, genotypes = genotypes, fc = 1.15)
mat <- chromatin.KO$results[,c(8:27,1,3)]
colnames(mat)[21] <- "gene.name"

```
Run k-means clustering on control (WT) group
```{r}
grp.idx <- WTgrp_kmeans(control_mat = mat[,1:10])
if(length(grp.idx$WT.idx1) != length(grp.idx$WT.idx2)){
message("Cluster size is not equal, therefore run same size k-means variation!")
    grp.idx <- WTgrp_kmeans_eqSize(control_mat = mat[,1:10])
}

```
Plot running average figure with wrapper function
```{r}
res7 <- overlap_wrapper(dat = mat, refseq = refseq, KO.idx = c(11:20),
                        WT.idx = c(1:10), WT1.idx = grp.idx$WT.idx1,
                        WT2.idx = grp.idx$WT.idx2, bin.size = 200,
                        shift.size = 40)
res7$plot

```
Using the DEGs list
```{r}
degs.dat <- read.table("../dat/DEGs/KO-WT_chromatin_RNA-seq.txt", sep = "\t",
                       stringsAsFactors=FALSE, header = TRUE, row.names = 1)
degs.dat <- degs.dat[degs.dat$FDR < 0.05,]
cat("Number of DEGs =",dim(degs.dat)[1],"\n\n")
degs.dat <- inner_join(degs.dat %>% rownames_to_column(var = "gene.name"),
                       refseq, by = "gene.name")

```
Generate scatter plots of log fold change vs gene length.
```{r}
Plot.Scatter(dat = degs.dat[,c("gene.name","logFC", "FDR", "gene.length")],
             log2FC = log2(1), comp.between = "")
Plot.Scatter(dat = degs.dat[,c("gene.name","logFC", "FDR", "gene.length")],
             log2FC = log2(1.2), comp.between = "")
degs.dat <- inner_join(x = chromatin.KO$results[chromatin.KO$results$gene %in%
                                                degs.dat$gene.name,c(8:27,1,3)],
                       y = degs.dat[,c("gene.name", "logFC")],
                       by = c("gene" = "gene.name"))
mat <- degs.dat[,c(1:21,23)]
colnames(mat)[21] <- "gene.name"

```
logFC from Degs from edgeR (Boxer et al.)
```{r}
colnames(mat)[22] <- "log2FoldChange"
res8 <- overlap_wrapper(dat = mat, refseq = refseq, KO.idx = c(11:20),
                        WT.idx = c(1:10), WT1.idx = grp.idx$WT.idx1,
                        WT2.idx = grp.idx$WT.idx2, bin.size = 40,
                        shift.size = 4, shrink_lfc = TRUE)
res8$plot
```
logFC from Degs from DESeq2
```{r}
mat <- degs.dat[,c(1:22)]
colnames(mat)[21] <- "gene.name"
res9 <- overlap_wrapper(dat = mat, refseq = refseq, KO.idx = c(11:20),
                        WT.idx = c(1:10), WT1.idx = grp.idx$WT.idx1,
                        WT2.idx = grp.idx$WT.idx2, bin.size = 40,
                        shift.size = 4, shrink_lfc = TRUE)
res9$plot
rm(dat, chromatin.KO, mat, grp.idx, degs.dat)

```
